src/MKL.jl

```diff
@@ -1,6 +1,41 @@
 module MKL
 
-using MKL_jll
+using Preferences
+using Libdl
+using LinearAlgebra
+
+# Choose an MKL path; taking an explicit preference as the first choice,
+# but if nothing is set as a preference, fall back to the default choice of `MKL_jll`.
+const mkl_path = something(
+    @load_preference("mkl_path", nothing),
+    "mkl_jll",
+)::String
+
+if lowercase(mkl_path) == "mkl_jll"
+    # Only load MKL_jll if we are suppoed to use it as the MKL source
+    # to avoid an unnecessary download of the (lazy) artifact.
+    import MKL_jll
+    const libmkl_rt = MKL_jll.libmkl_rt
+elseif lowercase(mkl_path) == "system"
+    # We expect the "system" MKL to already be loaded,
+    # or be on our linker search path.
+    libname = string("libmkl_rt", ".", Libdl.dlext)
+    const libmkl_rt = find_library(libname, [""])
+    libmkl_rt == "" && error("Couldn't find $libname. Try to specify the path to `libmkl_rt` explicitly.")
+else
+    # mkl_path should be a valid path to libmkl_rt.
+    const libmkl_rt = mkl_path
+    isfile(libmkl_rt) || error("Couldn't find MKL library at $libmkl_rt.")
```

Comment:
I'm not really a fan of errors at precompile time, especially since the function that most users would use to fix this (`set_mkl_path()`) is only available from a properly-precompiled package image.  I would instead raise a warning at precompile time here, and then another warning at `__init__()` time in the event of a clear misconfiguration like this.